jlox/app/src/main/java/dev/zxul767/lox/runtime/LoxList.java

```diff
@@ -3,62 +3,100 @@
+        new CallableSignature("at", new Parameter("index", "int")),
+        new CallableSignature("__getitem__", new Parameter("index", "int")),
+        new CallableSignature(
+            "set", Parameter.list("index", "int").add("value").build()
+        ),
+        new CallableSignature(
+            "__setitem__", Parameter.list("index", "int").add("value").build()
+        ),
+        new CallableSignature("clear", "nil"), new CallableSignature("pop")
+    );
+    signatures = new HashMap<>();
+    for (CallableSignature signature : members) {
+      signatures.put(signature.name, signature);
+    }
+  }
+  static Map<String, LoxCallable> buildConstructorMethod() {
+    Map<String, LoxCallable> methods = new HashMap<String, LoxCallable>();
+    methods.put(
+        "__init__",
+        new NativeBoundMethod(
+            signatures.get("__init__"), (self, args) -> { return null; }
+        )
+    );
+    return methods;
   }
 
   LoxList() {
     super(
         "list", /*superclass:*/ null,
-        /*methods:*/ new HashMap<String, LoxCallable>()
+        /*methods:*/ buildConstructorMethod()
     );
 
-    registerMethod("length", 0, (self, args) -> (double)self.list.size());
-    registerMethod("append", 1, (self, args) -> self.list.add(args[0]));
-    registerMethod("at", 1, (self, args) -> at(self, args[0]));
-    registerMethod("__getitem__", 1, (self, args) -> at(self, args[0]));
-    registerMethod("set", 2, (self, args) -> set(self, args[0], args[1]));
+    registerMethod("length", (self, args) -> length(self));
+    registerMethod("append", (self, args) -> add(self, args[0]));
+    registerMethod("at", (self, args) -> at(self, args[0]));
+    registerMethod("__getitem__", (self, args) -> at(self, args[0]));
+    registerMethod("set", (self, args) -> set(self, args[0], args[1]));
     registerMethod(
-        "__setitem__", 2, (self, args) -> chainable_set(self, args[0], args[1])
+        "__setitem__", (self, args) -> chainable_set(self, args[0], args[1])
     );
-    registerMethod("clear", 0, (self, args) -> clear(self));
-    registerMethod("pop", 0, (self, args) -> pop(self));
+    registerMethod("clear", (self, args) -> clear(self));
+    registerMethod("pop", (self, args) -> pop(self));
+  }
+
+  // TODO: Create a NativeClass subclass so we can hoist common methods and code
+  // between LoxString and LoxList
+  void
+  registerMethod(String name, NativeMethod<Object, Object> nativeFunction) {
+    CallableSignature signature = signatures.getOrDefault(name, null);
+    if (signature == null) {
+      System.err.println(
+          String.format("ERROR: failed to register method: %s", name)
+      );
+    } else {
+      this.methods.put(
+          signature.name, new NativeBoundMethod(signature, nativeFunction)
+      );
+    }
   }
 
-  void registerMethod(String name, int arity, Invoker<Object, Object> invoker) {
-    this.methods.put(name, new LoxListMethod(name, arity, invoker));
+  // constructor call
```

Comment:
We should add a comment that this is like `__new__` in Python, which is then followed (if present) by `__init__`